Transactional session documentation updates and cleanup (#5860)
Particular/docs.particular.net: nservicebus/transactional-session/index.md

    * This results is a user in the database, known as a phantom record, which is never announced to the rest of the system.
2. **Ghost message**: The controller publishes the `UserCreated` event first, then creates the user in the database. If a failure occurs between these two operations:
    * The `UserCreated` event is published, but the user is not created in the database.

>     * The rest of the system is notified about the creation of the user, but the record does not exist in the database. This inconsistency causes errors in the system, as parts of the system expect the user to exist in the database.

Reviewer: ```suggestion
    * The rest of the system is notified about the creation of the user, but the user record is never created. This inconsistency causes errors, as parts of the system expect the record to exist in the database.
```

In the context of a message handler, the [Outbox](/nservicebus/outbox) feature can mitigate this problem. However, such scenarios remain unsolved outside the context of a message handler.

A common technique to address this problem on the client side is to defer all operations to a message handler. This entails sending a message to create the user and publishing the `UserCreated`-event from within a message handler.
However, there are scenarios where this approach is not feasible:

> * Existing applications that want to introduce messaging already have quite some logic in controllers. Moving all that logic into dedicated message handlers requires significant effort, and might not be feasible from day one.

Reviewer: ```suggestion
* Existing applications are likely to have non-trivial processing logic in controllers. Moving it all into dedicated message handlers requires significant effort.
```

> * Existing logic in the controller might assume certain side effects to occur within the scope of the request (for example validation, notifications or error-handling) and not yet ready to fully embrace the asynchronous and fire&forget nature of offloading the work into message handlers.

Author: Brought back the `asynchronous and` part because I like driving this home ;)

> * Existing logic in the controller may assume specific side effects to occur within the scope of the request, e.g., validation, notifications, or error handling. The logic may, therefore, not be ready to fully embrace the asynchronous and fire&forget nature of offloading work into message handlers.

Reviewer: ```suggestion
* Processing logic in the controller may assume specific side effects to occur within the scope of the request, e.g., validation, notifications, or error handling. The logic may, therefore, not be ready for the asynchronous, fire-and-forget approach required in distributed messaging systems.
```

* There may be other scenarios in which it's not feasible to delay the database operation.

> The `TransactionalSession` feature solves this problem for messages sent/published outside the context of a message handler when combined with the [Outbox](/nservicebus/outbox) feature.

Reviewer: ```suggestion
The `TransactionalSession`, when combined with [Outbox](/nservicebus/outbox), solves this problem for messages sent and/or published outside of the context of a message handler.
```

Author: According to my grammar check "of the context of" is wrong. I will slightly tweak the suggestion

snippet: sending-transactional-session

The persistence-specific database session is accessible via the `transactionalSession.SynchronizedStorageSession` property or via dependency injection. See the [persistence-specific documentation](/nservicebus/transactional-session/persistences) for more details.

Once all the operations that are part of the atomic request have been executed, the session should be committed:

snippet: committing-transactional-session

Disposing the transactional session, without committing, will roll back any changes that were made.

> ### Advanced

Reviewer 2: ```suggestion
### Advanced settings
```

Reviewer 2: this could also be a `##` section?

Author: I placed it as a subsection to the usage settings and that's why it only says "advanced" in reference to "advanced usage"

#### Maximum commit duration

The maximum commit duration defaults to `TimeSpan.FromSeconds(15)`.

> The maximum commit duration limits the amount of time the transaction can attempt to commit the changes atomically. The value can be configured when opening the session.

Reviewer: ```suggestion
The maximum commit duration limits the amount of time it can take for a transaction to commit the changes. The value can be configured when opening the session.
```

snippet: configuring-timeout-transactional-session

> The maximum commit duration does not represent the total transaction time, but rather the actual time observed from the perspective of the control message used to settle the transaction outcome. The total transaction time observed might be longer, considering delays in the transport due to latency, delayed delivery, load on the input queue, endpoint concurrency limits, and more.

Reviewer: ```suggestion
The maximum commit duration does not represent the total transaction time, but rather the time it takes to complete the commit operation (as observed from the perspective of the control message). In practice, the observed total commit time might be longer due to delays in the transport caused by latency, delayed delivery, load on the input queue, endpoint concurrency limits, and more.
```

> The maximum commit duration is not the actual total transaction time from the perspective of the clock, but the actual time observed from the perspective of the control message that is used to settle the transaction outcome. The actual total transaction time observed might be longer, considering delays in the transport due to latency, delayed delivery, load on the input queue, endpoint concurrency limits and more.

Reviewer 3: ```suggestion
The maximum commit duration does not represent the total transaction time, but rather the actual time observed from the perspective of the control message used to settle the transaction outcome. The total transaction time observed might be longer, considering delays in the transport due to latency, delayed delivery, load on the input queue, endpoint concurrency limits, and more.
```

> When the control message arrives and the outbox record is not yet visible the following formula applies to delay the message (see [Phase 2](#how-it-works-phase-2)):

Reviewer 3: ```suggestion
When the control message is consumed, but the outbox record is not yet available in storage, the following formula is applied to delay the message (see [Phase 2](#how-it-works-phase-2)):
```

```text
CommitDelayIncrement = 2 * CommitDelayIncrement;
RemainingCommitDuration = RemainingCommitDuration - 
   (CommitDelayIncrement > RemainingCommitDuration ? RemainingCommitDuration : CommitDelayIncrement)
```

The default commit delay increment is set to `Timespan.FromSeconds(2)`and cannot be overridden.

#### Metadata

> It is possible to add custom headers to the control message that is used to settle the transaction outcome. Custom headers can be set by adding them to the metadata when opening the session.

Reviewer 4: Should we add information why would you ever want to do so? Like

```suggestion
It is possible to add custom headers to the control message that is used to settle the transaction outcome. Custom headers can be set by adding them to the metadata when opening the session. These headers can be access by a [custom behavior](https://docs.particular.net/nservicebus/pipeline/manipulate-with-behaviors#add-a-new-step) in the `TransportReceive` part of the pipeline.
```

Author: I was thinking about that and then I dismissed it because I felt that probably you never ever want to manipulate these control messages. Happy though to accept the suggestion

> It is possible to add custom headers to the control message that is used to settle the transaction outcome. Custom headers can be set by adding them to the metadata when opening the session. These headers can be accessed by a [custom behavior](/nservicebus/pipeline/manipulate-with-behaviors#add-a-new-step) in the `TransportReceive` part of the pipeline.

Author: ```suggestion
It is possible to add custom headers to the control message that is used to settle the transaction outcome. Custom headers can be set by adding them to the metadata when opening the session. These headers can be accessed by a [custom behavior](/nservicebus/pipeline/manipulate-with-behaviors.md#add-a-new-step) in the `TransportReceive` part of the pipeline.
```

> It is possible to add custom headers to the control message that is used to settle the transaction outcome. Custom headers can be set by adding them to the metadata when opening the session. These headers can be accessed by a [custom behavior](/nservicebus/pipeline/manipulate-with-behaviors.md#add-a-new-step) in the `TransportReceive` part of the pipeline.

Reviewer 2: would it be worthwhile to emphasize that the metadata values are not propagated to messages sent via the send/publish operations?

Author: I think that is implied enough by saying they are only applied to the control message

## Transaction consistency

> To guarantee atomic consistency across database and message operations, the transactional session requires the [Outbox](/nservicebus/outbox) to be enabled. Using the transactional session together with the outbox provides the strongest consistency guarantees and is therefore the recommended safe-by-default configuration.

Reviewer 3: ```suggestion
To guarantee atomic consistency across database and message operations, the transactional session requires the [Outbox](/nservicebus/outbox) to be enabled. This combination of features provides the strongest consistency guarantees and is, therefore, the recommended, safe-by-default configuration.
```

> NOTE: The outbox has to be [enabled explicitly](/nservicebus/outbox/#enabling-the-outbox) on the endpoint configuration.

Reviewer: ```suggestion
NOTE: Outbox has to be [enabled explicitly](/nservicebus/outbox/#enabling-the-outbox) on the endpoint configuration.
```

> With the Outbox disabled, database and message operations are not executed until the session is committed. All database operations share the same database transaction and are committed first. When the database operations are successful, the message operations are [batch-dispatched by the transport](/nservicebus/messaging/batched-dispatch.md). The message operations are not guaranteed to be atomic with the database changes. This might lead to phantom records or ghost messages in case of a failure during the commit phase.

Reviewer: ```suggestion
With the Outbox disabled, database and message operations are not applied until the session is committed. All database operations share the same database transaction and are committed first. When the database operations complete successfully, the message operations are [batch-dispatched by the transport](/nservicebus/messaging/batched-dispatch.md). The message operations and the database changes are not guaranteed to be atomic. This might lead to phantom records or ghost messages in case of a failure during the commit phase.
```

> With the Outbox disabled, database and message operations are not executed until the session is committed. All database operations share the same database transaction and are commit first. When the database operations were successful, the message operations are [batch-dispatched by the transport](/nservicebus/messaging/batched-dispatch.md). The message operations are not guaranteed to be atomic with the database changes. This might lead to phantom records or ghost messages when in case of a failure during the commit phase.

Reviewer 3: ```suggestion
With the Outbox disabled, database and message operations are not executed until the session is committed. All database operations share the same database transaction and are committed first. When the database operations are successful, the message operations are [batch-dispatched by the transport](/nservicebus/messaging/batched-dispatch.md). The message operations are not guaranteed to be atomic with the database changes. This might lead to phantom records or ghost messages in case of a failure during the commit phase.
```

4. The user can execute any required message operations using the transactional session.
5. The user can store any data using the persistence-specific session, which is accessible through the transactional session.
6. When all operations are registered, the user calls ´Commit´ on the transactional session.

> 7. A control message to settle the transaction is dispatched to the local queue. The control message is independent of the message operations and not stored in the outbox record.

Reviewer 3: ```suggestion
7. A control message to complete the transaction is dispatched to the local queue. The control message is independent of the message operations and is not stored in the outbox record.
```

> 7. A control message to complete the transaction is dispatched to the local queue. The control message is independent of the message operations and is not stored in the outbox record.

Reviewer 2: "to complete the transaction" is confusing since the transaction in this context is the storage transaction

```suggestion
7. A control message to trigger the dispatch mechanism is sent to the local queue. The control message is independent of the message operations and is not stored in the outbox record.
```

Author: We discussed it and are keeping the current wording

8. The message operations are converted and stored into an outbox record.
9. The transaction is committed, and the outbox record and business data modifications are stored atomically.

### Phase 2

The endpoint receives the control message and processes it as follows:

* Find the outbox record.

>   * If it exists, it hasn't been marked as dispatched, and there are pending operations, dispatch those operations, and set the outbox record as dispatched.

Reviewer: ```suggestion
  * If it exists, it hasn't been marked as dispatched, and there are pending operations than dispatch those operations, and set the outbox record as dispatched.
```

>   * If it doesn't exist, delay the processing of the control message.

Reviewer 2: ```suggestion
  * If it doesn't exist yet, delay the processing of the control message.
```

>   * If it doesn't exist, processing of the control message is delayed to a later point.

Reviewer: ```suggestion
  * If it doesn't exist, delay the processing of the control message.
```

## Failure scenarios

> The transactional session provides atomic store-and-send guarantees, similar to Outbox (except for incoming message de-duplication). That said, it cannot rely on the recoverability mechanism used by the Outbox, which uses retries to ensure outgoing messages are dispatched when failures occur. Instead, the control message is used to ensure that **exactly one** of these two outcomes occur:

Reviewer 2: ```suggestion
The transactional session provides atomic store-and-send guarantees, similar to Outbox (except for incoming message de-duplication). That said, it cannot rely on the recoverability mechanism used by the Outbox, which uses [retries](/nservicebus/recoverability) to ensure outgoing messages are dispatched when failures occur. Instead, the control message is used to ensure that **exactly one** of these two outcomes occur:
```

> The transactional session provides atomic store-and-send guarantees similar to Outbox (but no incoming message de-duplication). That said, it cannot rely on the recoverability mechanism used in the `Outbox` that ensures pushing the outgoing messages out in face of a failure. Instead, the previously mentioned control message is used to make sure that **exactly one** of these two outcomes happens:

Reviewer 3: ```suggestion
The transactional session provides atomic store-and-send guarantees, similar to Outbox (except for incoming message de-duplication). That said, it cannot rely on the recoverability mechanism used by the Outbox, which uses retries to ensure outgoing messages are dispatched when failures occur. Instead, the control message is used to ensure that **exactly one** of these two outcomes occur:
```

> * Transaction finishes with no visible side effects - when the control message stores the `OutboxRecord`
> * Transaction finishes with data being stored, and outgoing messages eventually sent - when the `Commit` path successfully stores the `OutboxRecord`

Reviewer 2: Should the happy path be described first?

```suggestion
* Transaction finishes with data being stored, and outgoing messages eventually sent - when the `Commit` path successfully stores the `OutboxRecord`
* Transaction finishes with no visible side effects - when the control message stores the `OutboxRecord`
```

> * Transaction finishes with data stored, and outgoing messages eventually sent - when the `Commit` path successfully stores the `OutboxRecord`

Reviewer 3: ```suggestion
* Transaction finishes with data stored, and outgoing messages are eventually sent - when the `Commit` path successfully stores the `OutboxRecord`
```

> * Transaction finishes with data stored, and outgoing messages are eventually sent - when the `Commit` path successfully stores the `OutboxRecord`

Reviewer: ```suggestion
* Transaction finishes with data being stored, and outgoing messages eventually sent - when the `Commit` path successfully stores the `OutboxRecord`
```

> Sending the control message first ensures that, eventually, the transaction will have an atomic outcome. If the `Commit` of the `OutboxRecord` succeeds, the control message will ensure the outgoing operations are sent out. If the `Commit` fails, the control message will, with respect to the [maximum commit duration](#usage-advanced-maximum-commit-duration) eventually be consumed, leaving no side effects.

Reviewer: ```suggestion
Sending the control message first ensures that, eventually, the transaction will have an atomic outcome. If the `Commit` of the `OutboxRecord` succeeds, the control message will ensure the outgoing operations are sent out. If the `Commit` fails, the control message will (after the [maximum commit duration](#usage-advanced-maximum-commit-duration) elapses) eventually be consumed, leaving no side effects.
```

> Sending the control message first ensures that eventually the transaction will have an atomic result. If the `Commit` of the `OutboxRecord` succeeds, the control message will make sure the outgoing operations are sent out. If the `Commit` fails, the control message will be eventually (after the [maximum commit duration](#usage-advanced-maximum-commit-duration) elapses) consumed, leaving no side effects.
> 
> When dispatching of the control message fails, the transactional session changes will be rolled-back and an error is raised to the user committing the session.

Reviewer 3: ```suggestion
Sending the control message first ensures that, eventually, the transaction will have an atomic outcome. If the `Commit` of the `OutboxRecord` succeeds, the control message will ensure the outgoing operations are sent out. If the `Commit` fails, the control message will [eventually](#usage-advanced-maximum-commit-duration) be consumed, leaving no side effects.

When dispatching the control message fails, the transactional session changes will be rolled back, and an error is raised to the user committing the session.
```

Author: Slightly twisted this suggestion

> When dispatching the control message fails, the transactional session changes will be rolled back, and an error is raised to the user committing the session.

Reviewer: ```suggestion
When dispatching the control message fails, the transactional session changes will be rolled back, and an error will be raised to the user committing the session.
```

## Limitations

> * The transactional session requires a full endpoint to send a control message to the local queue, and therefore cannot be used in send-only endpoints.

Reviewer: ```suggestion
* The transactional session cannot be used in send-only endpoints. A full endpoint is required to send a control message to the local queue.
```

> * Deduplication is guaranteed in Phase 2, but not in Phase 1. In Phase 2, the Outbox record ensures that the operations will never be dispatched more than once. However, in Phase 1, the session is always assigned a unique identifier and thus the operations in Phase 1 cannot be retried.

Reviewer 2: the outbox mechanism only marks all dispatched messages as dispatched after all have been sent. When dispatching 3 messages and the third one fails, the process will attempt to re-dispatch all 3 messages again, doesn't it?

Author: Removed the whole bullet after discussing it with @tmasternak and @SzymonPobiega since we agreed it doesn't convey much anyway

> * Deduplication is guaranteed in Phase 2, but not in Phase 1. In Phase 2, the outbox record ensures that the operations will never be dispatched more than once. However, during Phase 1, a unique ID is assigned to the session automatically and thus the operations in Phase 1 cannot be retried.

Reviewer: ```suggestion
* Deduplication is guaranteed in Phase 2, but not in Phase 1. In Phase 2, the Outbox record ensures that the operations will never be dispatched more than once. However, in Phase 1, the session is always assigned a unique identifier and thus the operations in Phase 1 cannot be retried.
```

> * The transactional session cannot be used in send-only endpoints. A full endpoint is required to send a control message to the local queue.
> * Deduplication is guaranteed in Phase 2, but not in Phase 1. In Phase 2, the Outbox record ensures that the operations will never be dispatched more than once. However, in Phase 1, the session is always assigned a unique identifier and thus the operations in Phase 1 cannot be retried.

Author: ```suggestion
* The transactional session cannot be used in send-only endpoints. A full endpoint is required to send a control message to the local queue.
```
